src/pymor/algorithms/chol_qr.py

```diff
@@ -0,0 +1,169 @@
+        repeat the algorithm until the check passes or `maxiter` is reached.
+    check_finite
+        This argument is passed down to |SciPy linalg| functions. Disabling may give a
+        performance gain, but may result in problems (crashes, non-termination) if the
+        inputs do contain infinities or NaNs.
+    copy
+        If `True`, create a copy of `A` instead of modifying `A` in-place.
+
+    Returns
+    -------
+    Q
+        The orthonormalized |VectorArray|.
+    R
+        The upper-triangular/trapezoidal matrix (if `compute_R` is `True`).
+    """
+    assert 0 <= offset <= len(A)
+    assert 0 < maxiter
+    assert orth_tol is None or 0 < orth_tol
+
+    if copy:
+        A = A.copy()
+
+    if offset == len(A):
+        return A, np.eye(len(A))
+    elif A.sup_norm().max() < 1e-16:
+        k = len(A)
+        del A[:]
+        R = np.empty([len(A), k])
+        return A, R
+
+    logger = getLogger('pymor.algorithms.chol_qr.shifted_chol_qr')
+
+    if maxiter == 1:
+        logger.warning('Single iteration shifted CholeskyQR can lead to poor orthogonality!')
+
+    B, X = np.split(A[offset:].inner(A, product=product), [offset], axis=1)
+    B = B.conj()
+
+    dtype = np.promote_types(X.dtype, np.float32)
+    eps = np.finfo(dtype).eps
+    if dtype == np.float32:
+        xtrmm, xtrtri = spla.blas.strmm, spla.lapack.strtri
+    elif dtype == np.float64:
+        xtrmm, xtrtri = spla.blas.dtrmm, spla.lapack.dtrtri
+    elif dtype == np.complex64:
+        xtrmm, xtrtri = spla.blas.ctrmm, spla.lapack.ctrtri
+    elif dtype == np.complex128:
+        xtrmm, xtrtri = spla.blas.ztrmm, spla.lapack.ztrtri
+
+    iter = 1
+    shift = None
+    eig = None
+    while iter <= maxiter:
+        shift = None
+        with logger.block(f'Iteration {iter}'):
+            # This will compute the Cholesky factor of the lower right block
+            # and keep applying shifts if it breaks down.
+            while True:
+                try:
+                    Rx = spla.cholesky(X - B@B.T, overwrite_a=True, check_finite=check_finite)
+                    break
+                except spla.LinAlgError:
+                    pass
+                logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+                if shift is None:
+                    m, n = A.dim, len(A[offset:])
+                    if product is None:
+                        shift = m*n+n*(n+1)
+                        XX = X
+                    else:
+                        if not eig:
+                            from pymor.algorithms.eigs import eigs
+                            eig = eigs(product, k=1)[0][0]
+
+                        shift = 2*m*np.sqrt(m*n)+n*(n+1)*np.abs(eig)
+                        XX = A[offset:].gramian(product=product)
+                    shift *= 11*eps*spla.eigh(XX, eigvals_only=True, subset_by_index=[n-1, n-1], driver='evr')[0]**2
```

Comment:
Wow, there was a massive mistake because A and X are used exatcly conversely in the paper and code. Of course, the shift only needs to be computed once. I've fixed this, resulting in yet another speedup!